File: tests/test_cards.py
import pytest
from gto.cards import Card, Deck, parse_card, parse_board, simplify_hand, hand_combos


class TestCard:
    def test_creation(self):
        c = Card("A", "s")
        assert c.rank == "A"
        assert c.suit == "s"
        assert c.value == 14

    def test_invalid_rank(self):
        with pytest.raises(ValueError):
            Card("X", "s")

    def test_invalid_suit(self):
        with pytest.raises(ValueError):
            Card("A", "x")

    def test_str(self):
        assert str(Card("K", "d")) == "Kd"

    def test_pretty(self):
        assert Card("A", "s").pretty() == "A\u2660"

    def test_ordering(self):
        assert Card("2", "s") < Card("A", "s")
        assert not Card("K", "h") < Card("Q", "d")

    def test_equality(self):
        assert Card("A", "s") == Card("A", "s")
        assert Card("A", "s") != Card("A", "h")

    def test_hashable(self):
        s = {Card("A", "s"), Card("A", "s"), Card("K", "h")}
        assert len(s) == 2


class TestParseCard:
    def test_basic(self):
        assert parse_card("As") == Card("A", "s")
        assert parse_card("Td") == Card("T", "d")

    def test_case_insensitive_suit(self):
        assert parse_card("AH") == Card("A", "h")

    def test_invalid(self):
        with pytest.raises(ValueError):
            parse_card("ABC")


class TestParseBoard:
    def test_flop(self):
        board = parse_board("AsKdQh")
        assert len(board) == 3
        assert board[0] == Card("A", "s")

    def test_with_spaces(self):
        board = parse_board("As Kd Qh")
        assert len(board) == 3

    def test_turn(self):
        board = parse_board("AsKdQh5c")
        assert len(board) == 4

    def test_river(self):
        board = parse_board("As Kd Qh 5c 2s")
        assert len(board) == 5


class TestDeck:
    def test_full_deck(self):
        d = Deck()
        assert len(d) == 52

    def test_exclude(self):
        excluded = [Card("A", "s"), Card("K", "h")]
        d = Deck(exclude=excluded)
        assert len(d) == 50

    def test_deal(self):
        d = Deck()
        cards = d.deal(5)
        assert len(cards) == 5
        assert len(d) == 47

    def test_deal_too_many(self):
        d = Deck()
        with pytest.raises(ValueError):
            d.deal(53)

    def test_shuffle(self):
        d = Deck()
        original = d.cards.copy()
        d.shuffle()
        assert len(d) == 52
        assert set(d.cards) == set(original)


class TestSimplifyHand:
    def test_pair(self):
        assert simplify_hand([Card("A", "s"), Card("A", "h")]) == "AA"

    def test_suited(self):
        assert simplify_hand([Card("A", "s"), Card("K", "s")]) == "AKs"

    def test_offsuit(self):
        assert simplify_hand([Card("A", "s"), Card("K", "h")]) == "AKo"

    def test_ordering(self):
        assert simplify_hand([Card("K", "s"), Card("A", "s")]) == "AKs"
        assert simplify_hand([Card("9", "h"), Card("T", "d")]) == "T9o"


class TestHandCombos:
    def test_pair(self):
        combos = hand_combos("AA")
        assert len(combos) == 6

    def test_suited(self):
        combos = hand_combos("AKs")
        assert len(combos) == 4
        for c1, c2 in combos:
            assert c1.suit == c2.suit

    def test_offsuit(self):
        combos = hand_combos("AKo")
        assert len(combos) == 12
        for c1, c2 in combos:
            assert c1.suit != c2.suit

    def test_specific(self):
        combos = hand_combos("AsKh")
        assert len(combos) == 1
        assert combos[0] == (Card("A", "s"), Card("K", "h"))

File: tests/test_cards.rs
use gto_cli::cards::*;
use gto_cli::error::GtoError;

#[test]
fn test_card_creation() {
    let c = Card::new(Rank::Ace, Suit::Spades);
    assert_eq!(c.rank, Rank::Ace);
    assert_eq!(c.suit, Suit::Spades);
    assert_eq!(c.value(), 14);
}

#[test]
fn test_invalid_rank() {
    assert!(Rank::from_char('X').is_err());
}

#[test]
fn test_invalid_suit() {
    assert!(Suit::from_char('x').is_err());
}

#[test]
fn test_card_str() {
    let c = Card::new(Rank::King, Suit::Diamonds);
    assert_eq!(format!("{}", c), "Kd");
}

#[test]
fn test_card_pretty() {
    let c = Card::new(Rank::Ace, Suit::Spades);
    assert_eq!(c.pretty(), "A\u{2660}");
}

#[test]
fn test_card_ordering() {
    let two = Card::new(Rank::Two, Suit::Spades);
    let ace = Card::new(Rank::Ace, Suit::Spades);
    assert!(two < ace);
    let king = Card::new(Rank::King, Suit::Hearts);
    let queen = Card::new(Rank::Queen, Suit::Diamonds);
    assert!(!(king < queen));
}

#[test]
fn test_card_equality() {
    let a1 = Card::new(Rank::Ace, Suit::Spades);
    let a2 = Card::new(Rank::Ace, Suit::Spades);
    let a3 = Card::new(Rank::Ace, Suit::Hearts);
    assert_eq!(a1, a2);
    assert_ne!(a1, a3);
}

#[test]
fn test_card_hashable() {
    use std::collections::HashSet;
    let mut s = HashSet::new();
    s.insert(Card::new(Rank::Ace, Suit::Spades));
    s.insert(Card::new(Rank::Ace, Suit::Spades)); // duplicate
    s.insert(Card::new(Rank::King, Suit::Hearts));
    assert_eq!(s.len(), 2);
}

#[test]
fn test_parse_card_basic() {
    assert_eq!(parse_card("As").unwrap(), Card::new(Rank::Ace, Suit::Spades));
    assert_eq!(parse_card("Td").unwrap(), Card::new(Rank::Ten, Suit::Diamonds));
}

#[test]
fn test_parse_card_case_insensitive_suit() {
    assert_eq!(parse_card("AH").unwrap(), Card::new(Rank::Ace, Suit::Hearts));
}

#[test]
fn test_parse_card_invalid() {
    assert!(parse_card("ABC").is_err());
}

#[test]
fn test_parse_board_flop() {
    let board = parse_board("AsKdQh").unwrap();
    assert_eq!(board.len(), 3);
    assert_eq!(board[0], Card::new(Rank::Ace, Suit::Spades));
}

#[test]
fn test_parse_board_with_spaces() {
    let board = parse_board("As Kd Qh").unwrap();
    assert_eq!(board.len(), 3);
}

#[test]
fn test_parse_board_turn() {
    let board = parse_board("AsKdQh5c").unwrap();
    assert_eq!(board.len(), 4);
}

#[test]
fn test_parse_board_river() {
    let board = parse_board("As Kd Qh 5c 2s").unwrap();
    assert_eq!(board.len(), 5);
}

#[test]
fn test_deck_full() {
    let d = Deck::new(None);
    assert_eq!(d.len(), 52);
}

#[test]
fn test_deck_exclude() {
    let excluded = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::King, Suit::Hearts),
    ];
    let d = Deck::new(Some(&excluded));
    assert_eq!(d.len(), 50);
}

#[test]
fn test_deck_deal() {
    let mut d = Deck::new(None);
    let cards = d.deal(5).unwrap();
    assert_eq!(cards.len(), 5);
    assert_eq!(d.len(), 47);
}

#[test]
fn test_deck_deal_too_many() {
    let mut d = Deck::new(None);
    assert!(d.deal(53).is_err());
}

#[test]
fn test_deck_shuffle() {
    let mut d = Deck::new(None);
    let original: std::collections::HashSet<Card> = d.cards.iter().copied().collect();
    d.shuffle();
    assert_eq!(d.len(), 52);
    let shuffled: std::collections::HashSet<Card> = d.cards.iter().copied().collect();
    assert_eq!(original, shuffled);
}

#[test]
fn test_simplify_pair() {
    let cards = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::Ace, Suit::Hearts),
    ];
    assert_eq!(simplify_hand(&cards).unwrap(), "AA");
}

#[test]
fn test_simplify_suited() {
    let cards = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::King, Suit::Spades),
    ];
    assert_eq!(simplify_hand(&cards).unwrap(), "AKs");
}

#[test]
fn test_simplify_offsuit() {
    let cards = vec![
        Card::new(Rank::Ace, Suit::Spades),
        Card::new(Rank::King, Suit::Hearts),
    ];
    assert_eq!(simplify_hand(&cards).unwrap(), "AKo");
}

#[test]
fn test_simplify_ordering() {
    let cards1 = vec![
        Card::new(Rank::King, Suit::Spades),
        Card::new(Rank::Ace, Suit::Spades),
    ];
    assert_eq!(simplify_hand(&cards1).unwrap(), "AKs");

    let cards2 = vec![
        Card::new(Rank::Nine, Suit::Hearts),
        Card::new(Rank::Ten, Suit::Diamonds),
    ];
    assert_eq!(simplify_hand(&cards2).unwrap(), "T9o");
}

#[test]
fn test_hand_combos_pair() {
    let combos = hand_combos("AA").unwrap();
    assert_eq!(combos.len(), 6);
}

#[test]
fn test_hand_combos_suited() {
    let combos = hand_combos("AKs").unwrap();
    assert_eq!(combos.len(), 4);
    for (c1, c2) in &combos {
        assert_eq!(c1.suit, c2.suit);
    }
}

#[test]
fn test_hand_combos_offsuit() {
    let combos = hand_combos("AKo").unwrap();
    assert_eq!(combos.len(), 12);
    for (c1, c2) in &combos {
        assert_ne!(c1.suit, c2.suit);
    }
}

#[test]
fn test_hand_combos_specific() {
    let combos = hand_combos("AsKh").unwrap();
    assert_eq!(combos.len(), 1);
    assert_eq!(combos[0].0, Card::new(Rank::Ace, Suit::Spades));
    assert_eq!(combos[0].1, Card::new(Rank::King, Suit::Hearts));
}
